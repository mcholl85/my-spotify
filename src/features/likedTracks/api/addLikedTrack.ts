import { useMutation, useQueryClient } from '@tanstack/react-query'
import api from '../../../lib/axios'
import { QUERY_LIKED_TRACKS } from '../../../constants/query.constants'

const addLikedTrack = async (id: string) => {
  await api.put(`/me/tracks?ids=${id}`)
}

export const useAddLikedTrack = () => {
  const queryClient = useQueryClient()

  return useMutation({
    mutationFn: addLikedTrack,
    onSuccess: () => {
      queryClient.invalidateQueries([QUERY_LIKED_TRACKS])
    },
  })
}
