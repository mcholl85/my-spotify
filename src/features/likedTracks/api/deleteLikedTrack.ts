import { useMutation, useQueryClient } from '@tanstack/react-query'
import api from '../../../lib/axios'
import { QUERY_LIKED_TRACKS } from '../../../constants/query.constants'
import { QueryData } from '../types/queryData'
import { changeLikeStatus } from '../utils/changeLikeStatus'

const deleteLikedTrack = async (id: string) => {
  await api.delete(`/me/tracks?ids=${id}`)
}

type useDeleteLikedTrackProps = {
  offset: number
}

export const useDeleteLikedTrack = ({ offset }: useDeleteLikedTrackProps) => {
  const queryClient = useQueryClient()

  return useMutation({
    mutationFn: deleteLikedTrack,
    onSuccess: (_, id) =>
      queryClient.setQueryData([QUERY_LIKED_TRACKS, offset], (oldData: QueryData) =>
        changeLikeStatus({ data: oldData, id, isLike: false }),
      ),
  })
}
