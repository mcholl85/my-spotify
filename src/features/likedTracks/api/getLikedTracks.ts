import api from '../../../lib/axios'
import { Track } from '../types/track'
import { useQuery } from '@tanstack/react-query'
import { QUERY_LIKED_TRACKS } from '../../../constants/query.constants'
import { DEFAULT_LIMIT, DEFAULT_OFFSET } from '../../../constants/api.constants'
import { setTrackList } from '../utils/setTrackList'

type GetLikedTracksParams = {
  limit?: number
  offset?: number
}

type useLikedTracksProps = GetLikedTracksParams

export type GetLikedTracksResponse = {
  href: string
  limit: number
  next: string
  offset: number
  previous: string
  total: number
  items: {
    added_at: Date
    track: Track
  }[]
}

const getLikedTracks = async ({
  limit = DEFAULT_LIMIT,
  offset = DEFAULT_OFFSET,
}: GetLikedTracksParams): Promise<GetLikedTracksResponse> => {
  return await (
    await api.get(`/me/tracks?limit=${limit}&offset=${offset}`)
  ).data
}

export const useLikedTracks = ({ limit, offset }: useLikedTracksProps) => {
  return useQuery({
    queryKey: [QUERY_LIKED_TRACKS, offset],
    queryFn: async () => {
      const data = await getLikedTracks({ limit, offset })
      return { total: data.total, tracks: setTrackList(data) }
    },
    retry: 0,
    refetchOnWindowFocus: false,
  })
}
