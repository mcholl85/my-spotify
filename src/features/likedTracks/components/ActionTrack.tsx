import { useEffect, useState } from 'react'
import { useAddLikedTrack } from '../api/addLikedTrack'
import { useDeleteLikedTrack } from '../api/deleteLikedTrack'
import { Button } from 'antd'

type ActionTrackProps = {
  id: string
}

export const ActionTrack = ({ id }: ActionTrackProps) => {
  const {
    mutate: addLikedTrack,
    isSuccess: addLikedTrackIsSuccess,
    isLoading: addLikedTrackIsLoading,
  } = useAddLikedTrack()
  const {
    mutate: removeLikedTrack,
    isSuccess: removeLikedTrackIsSuccess,
    isLoading: removeLikedTrackIsLoading,
  } = useDeleteLikedTrack()
  const [trackIsDelete, setTrackIsDelete] = useState(false)

  useEffect(() => {
    if (addLikedTrackIsSuccess) setTrackIsDelete(false)
  }, [addLikedTrackIsSuccess])

  useEffect(() => {
    if (removeLikedTrackIsSuccess) setTrackIsDelete(true)
  }, [removeLikedTrackIsLoading])

  return !trackIsDelete ? (
    <Button type='default' loading={removeLikedTrackIsLoading} onClick={() => removeLikedTrack(id)}>
      Enlever de Titres likés
    </Button>
  ) : (
    <Button type='primary' loading={addLikedTrackIsLoading} onClick={() => addLikedTrack(id)}>
      Ajouter aux Titres likés
    </Button>
  )
}
