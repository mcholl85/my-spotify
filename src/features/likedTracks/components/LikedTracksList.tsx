import Table, { ColumnsType } from 'antd/es/table'
import { Result, Space, Spin } from 'antd'
import { useLikedTracks } from '../api/getLikedTracks'
import { ActionTrack } from './ActionTrack'
import { setTrackList } from '../utils/setTrackList'
import { Column } from '../types/colum'

export const LikedTracksList = () => {
  const { data, isLoading, isError } = useLikedTracks({})
  const nbOfTracks = data?.items.length
  const columns: ColumnsType<Column> = [
    {
      title: 'Album',
      dataIndex: 'album',
      key: 'album',
      render: (_, record) => <img src={record.cover} alt={`${record.name} cover`} />,
    },
    {
      title: 'Titre',
      dataIndex: 'name',
      key: 'name',
    },
    {
      title: 'Action',
      key: 'action',
      render: (_, record) => <ActionTrack id={record.id} />,
    },
  ]

  if (isLoading)
    return (
      <Space size='large' className='container'>
        <Spin size='large' />
        <Space>Chargement en cours</Space>
      </Space>
    )
  if (isError) return <Result status='403' title='403' subTitle="Le token d'accès a expirée" />

  return (
    <Space direction='vertical' className='container'>
      <Result status='success' title={`Les ${nbOfTracks} musiques Spotify ont été chargées`} />
      <Table columns={columns} dataSource={setTrackList(data)} pagination={false} />
    </Space>
  )
}
