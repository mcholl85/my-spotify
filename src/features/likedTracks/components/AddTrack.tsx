import { useAddLikedTrack } from '../api/addLikedTrack'
import { DeleteTrack } from './DeleteTrack'

type AddTrackProps = {
  id: string
}

export const AddTrack = ({ id }: AddTrackProps) => {
  const { mutate: addLikedTrack, isSuccess } = useAddLikedTrack()

  if (isSuccess) return <DeleteTrack id={id} />

  return <button onClick={() => addLikedTrack(id)}>Ajouter aux Titres lik√©s</button>
}
