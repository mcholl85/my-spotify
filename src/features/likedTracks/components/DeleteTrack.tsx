import { useDeleteLikedTrack } from '../api/deleteLikedTrack'
import { AddTrack } from './AddTrack'

type DeleteTrackProps = {
  id: string
}

export const DeleteTrack = ({ id }: DeleteTrackProps) => {
  const { mutate: removeLikedTrack, isSuccess } = useDeleteLikedTrack()

  if (isSuccess) return <AddTrack id={id} />

  return <button onClick={() => removeLikedTrack(id)}>Enlever de Titres lik√©s</button>
}
