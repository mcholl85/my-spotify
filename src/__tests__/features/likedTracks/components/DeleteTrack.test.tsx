import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { QueryClient, QueryClientProvider } from '@tanstack/react-query'
import { PropsWithChildren } from 'react'
import { DeleteTrack } from '../../../../features/likedTracks/components/DeleteTrack'
import { useDeleteLikedTrack } from '../../../../features/likedTracks/api/deleteLikedTrack'

const queryClient = new QueryClient()
const wrapper = ({ children }: PropsWithChildren) => (
  <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
)
jest.mock('../../../../features/likedTracks/api/deleteLikedTrack', () => ({
  useDeleteLikedTrack: jest.fn(),
}))
const mockUseDeleteLikedTrack = useDeleteLikedTrack as jest.Mock

describe('Component DeleteTrack', () => {
  afterEach(() => {
    jest.clearAllMocks()
  })
  test('Matches DOM Snapshot', () => {
    mockUseDeleteLikedTrack.mockReturnValue({ mutate: jest.fn() })
    const { asFragment } = render(<DeleteTrack id={'123'} offset={50} />, { wrapper })

    expect(asFragment()).toMatchSnapshot()
  })
  test('should call useAddLikedTrack with the correct arguments', async () => {
    const mockedMutationFn = jest.fn()
    mockUseDeleteLikedTrack.mockReturnValue({
      mutate: mockedMutationFn,
      isLoading: false,
    })

    render(<DeleteTrack id={'123'} offset={0} />, { wrapper })

    await userEvent.click(screen.getByRole('button'))

    expect(mockedMutationFn).toBeCalledTimes(1)
    expect(mockedMutationFn).toBeCalledWith('123')
  })
  it('should display loading state when isLoading is true', () => {
    mockUseDeleteLikedTrack.mockReturnValue({
      mutate: jest.fn(),
      isLoading: true,
    })

    render(<DeleteTrack id='123' offset={0} />, { wrapper })

    expect(screen.getByRole('button')).toHaveClass('ant-btn-loading')
  })
  it('should not display loading state when isLoading is false', () => {
    mockUseDeleteLikedTrack.mockReturnValue({
      mutate: jest.fn(),
      isLoading: false,
    })

    render(<DeleteTrack id='123' offset={0} />, { wrapper })

    expect(screen.getByRole('button')).not.toHaveClass('ant-btn-loading')
  })
})
