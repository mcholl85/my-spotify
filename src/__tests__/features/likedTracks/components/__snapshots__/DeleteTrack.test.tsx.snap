// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component DeleteTrack Matches DOM Snapshot 1`] = `
<DocumentFragment>
  <button
    class="ant-btn css-dev-only-do-not-override-12jzuas ant-btn-default"
    type="button"
  >
    <span>
      Enlever de Titres lik√©s
    </span>
  </button>
</DocumentFragment>
`;
